tests/drivers/adc/adc_emul/src/main.c

```diff
@@ -0,0 +1,625 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/adc.h>
+#include <drivers/adc/adc_emul.h>
+#include <zephyr.h>
+#include <ztest.h>
+
+#define ADC_DEVICE_NAME		DT_LABEL(DT_INST(0, zephyr_adc_emul))
+#define ADC_REF_INTERNAL_MV	DT_PROP(DT_INST(0, zephyr_adc_emul), ref_internal_mv)
+#define ADC_REF_EXTERNAL1_MV	DT_PROP(DT_INST(0, zephyr_adc_emul), ref_external1_mv)
+#define ADC_RESOLUTION		14
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+#define ADC_1ST_CHANNEL_ID	0
+#define ADC_2ND_CHANNEL_ID	1
+
+#define INVALID_ADC_VALUE	SHRT_MIN
+/* Raw to milivolt conversion doesn't handle rounding */
+#define MV_OUTPUT_EPS		2
+#define SEQUENCE_STEP		100
+
+#define BUFFER_SIZE  6
+static ZTEST_BMEM int16_t m_sample_buffer[BUFFER_SIZE];
+
+const struct device *get_adc_device(void)
+{
+	return device_get_binding(ADC_DEVICE_NAME);
+}
+
+/*
+ * Setup channel with specific reference and gain
+ */
+static void channel_setup(const struct device *adc_dev, enum adc_reference ref,
+			  enum adc_gain gain, int channel)
+{
+	int ret;
+	struct adc_channel_cfg channel_cfg = {
+		.gain             = gain,
+		.reference        = ref,
+		.acquisition_time = ADC_ACQUISITION_TIME,
+		.channel_id       = channel,
+	};
+
+	ret = adc_channel_setup(adc_dev, &channel_cfg);
+	zassert_ok(ret, "Setting up of the %d channel failed with code %d",
+		   channel, ret);
+}
+
+/*
+ * Check if samples for specific channel are correct. It can check samples
+ * with arithmetic sequence with common difference.
+ */
```

Comment:
Please comment all args

Reply:
I have added doxygen style comments to all functions in this file (which include arguments description)